Add simple partials support

grego opened the pull request:
Adds very simple partials.
`{{>partial.html}}` now includes the file `partial.html`. All the used partials are stored in a cache, so they are never loaded twice and no copying of the source is required after it is loaded, the `Block`s in the template just reference the source from the partials as necessary.

In addition, a function `Templates::from_folder` is provided, which parses all the `.html` files in the given directory and stores them inside a (wrapped) hash map, making them accessible via their filename.
No file is ever loaded more than once and the source is never copied between templates.

The performance lead seems to be maintained with this PR, even when using partials, as they provide no overhead after parsing:
```
test pa_partials_ramhorns   ... bench:         216 ns/iter (+/- 4)
test pb_partials_wearte     ... bench:         236 ns/iter (+/- 7)
test pc_partials_askama     ... bench:         546 ns/iter (+/- 63)
test pd_partials_mustache   ... bench:       1,906 ns/iter (+/- 1,447)
test pe_partials_handlebars ... bench:       7,770 ns/iter (+/- 6,564)
```

maciejhirsz commented on ramhorns/src/template/parse.rs:
```diff
@@ -96,8 +100,23 @@ impl<'tpl> Template<'tpl> {
+                            Tag::Partial => {
+                                let path = dir.join(name);
+                                if !partials.contains_key(name) {
+                                    let template = Template::load(
```
I'm not sure I'm okay with this. Loading resources from fs should not be done within the parser, this could be extra dangerous if templates are user provided as we don't sanitize names here, so if you ever were to expose this on a web server to users, it could leak sensitive data by the virtue of doing `{{>../../../somefile}}`.

maciejhirsz replied:
A solution that just occured to me would be to use a generic here. Create a trait called `Partials` with a single method on it:

```rust
fn get_partial_blocks(&mut self, name: &str) -> Result<&[Block], Error>;
```

You can then have 2 separate implementations for it:

1. Always return an `Error` for when you parse a template from a `String` or `&str`.
2. Existing behaviour: load partials from disc with a local cache when loading a template from file.

There should be some sanitization here so that you cannot go up in the folder structure.

maciejhirsz commented on ramhorns/src/template/mod.rs:
```diff
@@ -37,14 +38,33 @@ pub struct Template<'tpl> {
     source: Cow<'tpl, str>,
+
+    /// Partials used in this template
+    partials: Option<Partials<'tpl>>,
```
We don't really need the entire `HashMap` at this point, we only care about having all the sources so that the compiler can make sure our `blocks` are always in memory, so a `Vec<Cow<'tpl, str>>` should be sufficient here. In theory, this could be rolled in with `source` to save us 3 words on the struct, at the cost of always doing an allocation, not sure that's better. Also reducing the `HashMap` to a `Vec` is extra work.

grego commented:
I implemented the `Partials` trait, as you suggested. However, the `get_pratial` method currently returns a reference to a `Template`, as the `capacity_hint` field is also required.
When loading a template from a file or from folder, the path of a partial is (https://doc.rust-lang.org/std/fs/fn.canonicalize.html)[canonicalized] and checked whether located within the same folder, otherwise rejected.
Moreover, I moved the `tail` filed into `blocks` to make this implementation more clean and terse, but if there is a particular need for it to be in a separate field, it should pose no problem to bring it back.
Lastly, I kept the `source` field and changed the type of the `partials` filed to `Vec<Cow<'tpl, str>>`. Merging those two would require to handle the sources differently and I'm not sure how to do that efficiently (also every template would need to make an allocation).

maciejhirsz commented:
I believe I put the tail in separately to avoid branches on looping, but since tag is already checked anyway that should be fine, so good catch! I'll review later, but just glimpsing over this looks good to me.

maciejhirsz commented:
Will make a release later today :)
